Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1332 @@
+  <hr>
+</details>
+
+### What's next?
+
+In this section, we've learned how to make a simple parsing combinator library. There are still various improvements that can be done. To name a few of them, we can implement more instances, such as `Semigroup` and `Monoid`, `IsString`, `MonadFail`, `MonadPlus`, among others. Error messages can be improved by better keeping track of location (such as line and column), supporting more types such as `Text` and `ByteString`, pretty-printing error messages, etc.
+
+We will conclude this part here, and if you are interested, proceed to the next part where Megaparsec, an industrial-strength parser combinator library, will be introduced with a more practical application than our simple examples in this section.
+
+## Megaparsec tutorial
+
+<!--
+* explain why we chose megaparsec
+* parse S-exps
+* make a small format for data parsing, no programming constructs
+  * equivalent to JSON: objects, lists, null, numbers, booleans, strings
+-->
+
+We will now take a look at a practical application with parser combinators: parsing S-expressions. Hopefully, such examples will give you a solid foundation for creating parsers for your grammars.
+
+Megaparsec is the go-to parser for industrial projects, so we will also use it in our tutorial. If you wish to see how it compares to other parser combinator libraries in Haskell, you can read its [`README.md`](https://github.com/mrkkrp/megaparsec#comparison-with-other-solutions).
+
+Keep in mind you should choose the parser library that better suits your needs. If you plan to parse machine-written data, for example, then Attoparsec may be a better choice. Its interface is similar to Megaparsec's, so the knowledge you will learn here will also be useful.
+
+Without any further ado, we will now present how to use Megaparsec to parse some [`S-expressions`](https://en.wikipedia.org/wiki/S-expressions).
+
+### Representing S-expressions
+
+To begin, create a file called `SExp.hs`. Also, make sure to get the [`megaparsec`](https://hackage.haskell.org/package/megaparsec) package.
+
+We will use the following data structure to represent various S-expressions. Gradually, we will write a parser that represents each of these cases.
+
+```hs
+module SExp where
+
+import Data.Void (Void)
+import Text.Megaparsec
+import Text.Megaparsec.Char
+
+newtype Identifier = Identifier
+  { getId :: String
+  } deriving (Show)
+
+data SExp
+  = SSExp    SExp [SExp]  -- (foo 0 "hello" bar), (bar (baz 1)), (foo)
+  | SInteger Integer  -- 42
+  | SString  String  -- "hello, world"
+  | SBool    Bool  -- false, true
+  | SId      Identifier  -- foo
+  deriving (Show)
+```
+
+We also need an alias for the type which `megaparsec` uses.
+
+```hs
+type Parser = Parsec
+  Void  -- The type for custom error messages. We have none, so use `Void`.
+  String  -- The input stream type. Let's use `String` for now, but for better performance, you might want to use `Text` or `ByteString`.
```

Comment:
I recommend splitting this line on 2 by hand, because otherwise it will be wrapped and wrapper doesn't care about comment indentation.

Reply:
I does go off-window with horizontal scrolling in Github preview.

Reply:
Good point. I've split them so it's shorter now.